libgnucash/engine/gnc-date.cpp

```diff
@@ -126,61 +126,55 @@ gnc_localtime_r (const time64 *secs, struct tm* time)
 }
 
 static void
-normalize_time_component (int *inner, int *outer, unsigned int divisor,
-                          int base)
+normalize_time_component (int *inner, int *outer, int divisor, int base)
 {
-     while (*inner < base)
-     {
-          --(*outer);
-          *inner += divisor;
-     }
-     while (*inner > static_cast<gint>(divisor))
-     {
-          ++(*outer);
-          *inner -= divisor;
-     }
+    if (*inner < base)
+    {
+        do
+        {
+            --(*outer);
+            *inner += divisor;
+        } while (*inner < base) ;
+        return;
+    }
+    while (*inner >= divisor + base)
+    {
+        ++(*outer);
+        *inner -= divisor;
+    }
 }
-
-static void
-normalize_month(int *month, int *year)
+  
+static int
+days_in_month (struct tm* time)
 {
-    ++(*month);
-    normalize_time_component(month, year, 12, 1);
-    --(*month);
+    normalize_time_component (&(time->tm_mon), &(time->tm_year), 12, 0);
+    return gnc_date_get_last_mday (time->tm_mon, time->tm_year + 1900);
 }
-
+  
 static void
 normalize_struct_tm (struct tm* time)
 {
-     gint year = time->tm_year + 1900;
-     gint last_day;
-
-     /* Gregorian_date throws if it gets an out-of-range year
-      * so clamp year into gregorian_date's range.
-      */
-     if (year < 1400) year += 1400;
-     if (year > 9999) year %= 10000;
```

Comment:
I think its bogus but, if you want it, I suggest using something more like:

  ```
if (year < 1400) year = 1400;
  else if (year > 9999) year = 9999;
```

Reply:
Neither corrects the likely mistakes of a 2-digit or 5-digit year, and it seems unlikely that anyone would be making book keeping entries in either century. Arguably the first should add something more than 1400.

Reply:
I agree.  The original code is just wrong, so I removed it.  Limiting the year range here, may mask a problem.  If we do, we should do so after we've normalized - not before.  Should we assert?  Show there be similar year range checks in `gnc-option-date.cpp:normalize_reldate_tm()`?

Reply:
> Limiting the year range here, may mask a problem.

What part of `Gregorian_date throws if it gets an out-of-range year` eludes your understanding?

> we should do so after we've normalized - not before

That would make sense only if there was a chance that the precondition fails because of an edge case. Nobody's using GnuCash to handle accounting events in either the 14th or 100th centuries. 

> Should we assert?

Asserts only work in debug builds. Dates come from user input in release builds.

>  Show there be similar year range checks in gnc-option-date.cpp:normalize_reldate_tm()?

No, because it's not part of the constructor for GncDateTime so it's not passing its result to boost::date_time.

Reply:
> > Limiting the year range here, may mask a problem.
> 
> What part of `Gregorian_date throws if it gets an out-of-range year` eludes your understanding?

I'm fine with `gregorian_date` throwing the error when the date is out-of-range.  I would prefer the `out-of-range year` be flagged on entry.

Since normalize_struct_tm() doesn't know why the year it receives is out-of-range, adjusting the year to make it be within range may make it more difficult to find the issue that caused the year to be out-of-range in the first place.

> 
> > we should do so after we've normalized - not before
> 
> That would make sense only if there was a chance that the precondition fails because of an edge case. 

There is a chance.

> Nobody's using GnuCash to handle accounting events in either the 14th or 100th centuries.

I agree.  So, why should normalize_struct_tm() need to check?

> 
> > Should we assert?
> 
> Asserts only work in debug builds. Dates come from user input in release builds.

Should we throw or return an error?

> 
> > Show there be similar year range checks in gnc-option-date.cpp:normalize_reldate_tm()?
> 
> No, because it's not part of the constructor for GncDateTime so it's not passing its result to boost::date_time.

So, we'll allow some dates to be out-of-range but not others.  An idiot could ask to see transactions  pre-14th or post-100th, but couldn't post them.

Reply:
I've reimplemented a guard in `normalize_struct_tm` that will set an date with a year earlier than 1400 to January 1, 1400 00:00:00 and any date with a year greater than  9999 to December 31, 9999 23:59:59.  

On an aside, there are 3 additional fields which are not being _normalized_: `tm_wday`, `tm_yday`, `tm_isdst`

Reply:
`mtkime` and `timegm` (sorry, I scrambled it with `localtime`; that emits a struct tm, it doesn't consume one) are specified to ignore those fields, as is boost::date_time's posix_time constructor.